File: Plugins/AM_PBFSpectraFileGen_PlugIn/clsAnalysisResourcesPBFGenerator.vb
'*********************************************************************************************************
' Written by Matthew Monroe for the US Department of Energy 
' Pacific Northwest National Laboratory, Richland, WA
' Created 07/16/2014
'
'*********************************************************************************************************

Option Strict On

Imports AnalysisManagerBase

Public Class clsAnalysisResourcesPBFGenerator
    Inherits clsAnalysisResources
    
    Public Overrides Function GetResources() As AnalysisManagerBase.IJobParams.CloseOutType

        ' Retrieve shared resources, including the JobParameters file from the previous job step
        Dim result = GetSharedResources()
        If result <> IJobParams.CloseOutType.CLOSEOUT_SUCCESS Then
            Return result
        End If

        If Not RetrieveInstrumentData() Then
            Return IJobParams.CloseOutType.CLOSEOUT_FILE_NOT_FOUND
        End If

        Return IJobParams.CloseOutType.CLOSEOUT_SUCCESS

    End Function

    Protected Function RetrieveInstrumentData() As Boolean

        Dim currentTask As String = "Initializing"

        Try

            Dim rawDataType As String = m_jobParams.GetJobParameter("RawDataType", "")
            Dim eRawDataType = GetRawDataType(rawDataType)

            If eRawDataType = eRawDataTypeConstants.ThermoRawFile Then
                m_jobParams.AddResultFileExtensionToSkip(DOT_RAW_EXTENSION)
            Else
                m_message = "PbfGen presently only supports Thermo .Raw files"
                Return False
            End If

            currentTask = "Retrieve intrument data"

            ' Retrieve the instrument data file
            If Not RetrieveSpectra(rawDataType) Then
                If String.IsNullOrEmpty(m_message) Then
                    m_message = "Error retrieving instrument data file"
                End If

                clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, "clsAnalysisResourcesPBFGenerator.GetResources: " & m_message)
                Return False
            End If

            currentTask = "Process MyEMSL Download Queue"

            If Not MyBase.ProcessMyEMSLDownloadQueue(m_WorkingDir, MyEMSLReader.Downloader.DownloadFolderLayout.FlatNoSubfolders) Then
                Return False
            End If

            Threading.Thread.Sleep(500)

            Return True

        Catch ex As Exception
            m_message = "Exception in RetrieveInstrumentData: " & ex.Message
            clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, m_message & "; task = " & currentTask & "; " & clsGlobal.GetExceptionStackTrace(ex))
            Return False
        End Try

    End Function

End Class

File: Plugins/AM_PBFSpectraFileGen_PlugIn/clsAnalysisResourcesPBFGenerator.cs
//*********************************************************************************************************
// Written by Matthew Monroe for the US Department of Energy
// Pacific Northwest National Laboratory, Richland, WA
// Created 07/16/2014
//
//*********************************************************************************************************

using AnalysisManagerBase;
using System;

namespace AnalysisManagerPBFGenerator
{

    /// <summary>
    /// Retrieve resources for the PBF Generator plugin
    /// </summary>
    public class clsAnalysisResourcesPBFGenerator : clsAnalysisResources
    {
        public override CloseOutType GetResources()
        {
            // Retrieve shared resources, including the JobParameters file from the previous job step
            var result = GetSharedResources();
            if (result != CloseOutType.CLOSEOUT_SUCCESS)
            {
                return result;
            }

            if (!RetrieveInstrumentData())
            {
                return CloseOutType.CLOSEOUT_FILE_NOT_FOUND;
            }

            return CloseOutType.CLOSEOUT_SUCCESS;
        }

        protected bool RetrieveInstrumentData()
        {
            var currentTask = "Initializing";

            try
            {
                var rawDataType = mJobParams.GetJobParameter("RawDataType", "");
                var eRawDataType = GetRawDataType(rawDataType);

                if (eRawDataType == eRawDataTypeConstants.ThermoRawFile)
                {
                    mJobParams.AddResultFileExtensionToSkip(DOT_RAW_EXTENSION);
                }
                else
                {
                    mMessage = "PbfGen presently only supports Thermo .Raw files";
                    return false;
                }

                currentTask = "Retrieve intrument data";

                // Retrieve the instrument data file
                if (!FileSearch.RetrieveSpectra(rawDataType))
                {
                    if (string.IsNullOrEmpty(mMessage))
                    {
                        mMessage = "Error retrieving instrument data file";
                    }

                    LogError("clsAnalysisResourcesPBFGenerator.GetResources: " + mMessage);
                    return false;
                }

                currentTask = "Process MyEMSL Download Queue";

                var success = ProcessMyEMSLDownloadQueue(mWorkDir, MyEMSLReader.Downloader.DownloadLayout.FlatNoSubdirectories);
                return success;
            }
            catch (Exception ex)
            {
                mMessage = "Exception in RetrieveInstrumentData: " + ex.Message;
                LogError(mMessage + "; task = " + currentTask + "; " + clsGlobal.GetExceptionStackTrace(ex));
                return false;
            }
        }
    }
}
